include/unifex/detail/debug_async_scope.hpp

```diff
@@ -0,0 +1,247 @@
+
+  template <typename... Values>
+  void set_value(Values&&... values) noexcept {
+    op_->complete([&](Receiver&& receiver) noexcept {
+      UNIFEX_TRY {
+        unifex::set_value(
+            std::move(receiver), static_cast<Values&&>(values)...);
+      }
+      UNIFEX_CATCH(...) {
+        unifex::set_error(std::move(receiver), std::current_exception());
+      }
+    });
+  }
+
+  template <typename E>
+  void set_error(E&& e) noexcept {
+    op_->complete([&e](Receiver&& receiver) noexcept {
+      unifex::set_error(std::move(receiver), static_cast<E&&>(e));
+    });
+  }
+
+  void set_done() noexcept { op_->complete(unifex::set_done); }
+};
+
+template <typename Sender, typename Receiver>
+struct _operation_impl final {
+  struct type;
+};
+
+template <typename Sender, typename Receiver>
+using debug_op =
+    typename _operation_impl<Sender, remove_cvref_t<Receiver>>::type;
+
+template <typename Sender, typename Receiver>
+struct _operation_impl<Sender, Receiver>::type final
+  : _operation<Receiver>::type {
+  using base_op_t = typename _operation<Receiver>::type;
+  using receiver_t = typename _receiver<Receiver>::type;
+
+  template <typename Sender2, typename Receiver2>
+  type(debug_op_list* ops, Sender2&& sender, Receiver2&& receiver) noexcept(
+      std::is_nothrow_constructible_v<
+          base_op_t,
+          const std::type_info&,
+          debug_op_list*,
+          Receiver2>&& is_nothrow_connectable_v<Sender2, receiver_t>)
+    : base_op_t(
+          typeid(connect_result_t<Sender, Receiver>),
+          ops,
+          static_cast<Receiver2&&>(receiver))
+    , op_(unifex::connect(static_cast<Sender2&&>(sender), receiver_t{this})) {}
+
+  type(type&&) = delete;
+
+  friend void tag_invoke(tag_t<start>, type& op) noexcept {
+    op.ops_->register_debug_operation(&op);
+    unifex::start(op.op_);
+  }
+
+private:
+  using op_t = connect_result_t<Sender, receiver_t>;
+  op_t op_;
+};
+
+template <typename Sender>
+struct _debug_scope_sender final {
+  struct type;
+};
+
+template <typename Sender>
+struct _debug_scope_sender<Sender>::type final {
+  template <
+      template <typename...>
+      typename Variant,
+      template <typename...>
+      typename Tuple>
+  using value_types = sender_value_types_t<Sender, Variant, Tuple>;
+
+  template <template <typename...> typename Variant>
+  using error_types = typename concat_type_lists_unique_t<
+      sender_error_types_t<Sender, type_list>,
+      type_list<std::exception_ptr>>::template apply<Variant>;
+
+  static constexpr bool sends_done = sender_traits<Sender>::sends_done;
+
+  template <typename Sender2>
+  type(Sender2&& sender, debug_op_list* ops) noexcept(
+      std::is_nothrow_constructible_v<Sender, Sender2>)
+    : ops_(ops)
+    , sender_(static_cast<Sender2&&>(sender)) {}
+
+  template(typename Self, typename Receiver)          //
+      (requires same_as<type, remove_cvref_t<Self>>)  //
+      friend auto tag_invoke(
```

Comment:
Might help compile time to say `debug_op<Sender, Receiver>` here.
```suggestion
      friend debug_op<Sender, Receiver> tag_invoke(
```